src/aind_behavior_video_transformation/etl.py

```diff
@@ -64,6 +73,37 @@ class BehaviorVideoJob(GenericEtl[BehaviorVideoJobSettings]):
     run_job() -> JobResponse
     """
 
+    def _run_compression(
+        self,
+        convert_video_args: list[tuple[Path, Path, tuple[str, str] | None]],
+    ) -> None:
+        """
+        Runs CompressionRequests at the specified paths.
+        """
+
+        if self.job_settings.parallel_compression:
+            # ProcessPool implementation
+            num_jobs = len(convert_video_args)
+            with ProcessPoolExecutor(max_workers=num_jobs) as executor:
+                jobs = [
+                    executor.submit(
+                        convert_video,
+                        *params,
+                        self.job_settings.ffmpeg_thread_cnt,
+                    )
+                    for params in convert_video_args
+                ]
+                for job in as_completed(jobs):
+                    result = job.result()
+                    logging.info("FFmpeg job completed:", result)
```

Comment:
This is the output of a working call (same input directory as test_run_job_with_data_structure):
"""
2024-11-07 11:04:40,762 - INFO - FFmpeg job completed: /private/var/folders/jp/tzg4yhb959s7g46rlqjjl2qh0000gp/T/tmpxq15v823/camera2/clip.mp4
2024-11-07 11:04:40,792 - INFO - FFmpeg job completed: /private/var/folders/jp/tzg4yhb959s7g46rlqjjl2qh0000gp/T/tmpxq15v823/camera1/clip.mp4
"""

This is the output if I set the requested compression to:
"""
faulty_req = CompressionRequest(
        compression_enum=CompressionEnum.USER_DEFINED,
        user_ffmpeg_input_options = "invalid user input args",
        user_ffmpeg_output_options = "invalid user output args",
)
"""

Nothing is logged and I believe stderr is forwarded to stdout:
"""
[AVFormatContext @ 0x11df07720] Unable to choose an output format for 'invalid'; use a standard extension for the filename or specify the format manually.
[AVFormatContext @ 0x12e006c50] [out#0 @ 0x11df12660] Unable to choose an output format for 'invalid'; use a standard extension for the filename or specify the format manually.
Error initializing the muxer for invalid: Invalid argument
Error opening output file invalid.
[out#0 @ 0x12e011b70] Error opening output files: Invalid argument
Error initializing the muxer for invalid: Invalid argument
Error opening output file invalid.
Error opening output files: Invalid argument
concurrent.futures.process._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/Users/jonathan.wong/miniconda3/envs/compression/lib/python3.11/concurrent/futures/process.py", line 261, in _process_worker
    r = call_item.fn(*call_item.args, **call_item.kwargs)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jonathan.wong/Projects/aind-behavior-video-transformation/src/aind_behavior_video_transformation/transform_videos.py", line 242, in convert_video
    subprocess.run(ffmpeg_command, check=True)
  File "/Users/jonathan.wong/miniconda3/envs/compression/lib/python3.11/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['ffmpeg', '-y', '-v', 'warning', '-hide_banner', 'invalid', 'user', 'input', 'args', '-i', '/private/var/folders/jp/tzg4yhb959s7g46rlqjjl2qh0000gp/T/tmp7ogiojea/camera2/clip.mp4', 'invalid', 'user', 'output', 'args', '/private/var/folders/jp/tzg4yhb959s7g46rlqjjl2qh0000gp/T/tmpxj4mbyd_/camera2/clip.mp4']' returned non-zero exit status 234.
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/jonathan.wong/Projects/aind-behavior-video-transformation/src/aind_behavior_video_transformation/etl.py", line 234, in <module>
    job_response = job.run_job()
                   ^^^^^^^^^^^^^
  File "/Users/jonathan.wong/Projects/aind-behavior-video-transformation/src/aind_behavior_video_transformation/etl.py", line 140, in run_job
    self._run_compression(convert_video_args)
  File "/Users/jonathan.wong/Projects/aind-behavior-video-transformation/src/aind_behavior_video_transformation/etl.py", line 97, in _run_compression
    result = job.result()
             ^^^^^^^^^^^^
  File "/Users/jonathan.wong/miniconda3/envs/compression/lib/python3.11/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/jonathan.wong/miniconda3/envs/compression/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
subprocess.CalledProcessError: Command '['ffmpeg', '-y', '-v', 'warning', '-hide_banner', 'invalid', 'user', 'input', 'args', '-i', '/private/var/folders/jp/tzg4yhb959s7g46rlqjjl2qh0000gp/T/tmp7ogiojea/camera2/clip.mp4', 'invalid', 'user', 'output', 'args', '/private/var/folders/jp/tzg4yhb959s7g46rlqjjl2qh0000gp/T/tmpxj4mbyd_/camera2/clip.mp4']' returned non-zero exit status 234.
"""

Is this garbled? It appears like this returns an error for any failed process.

Reply:
It looks like the two outputs are being interleaved in a way that makes it hard to follow, for example:
```
[AVFormatContext @ 0x11df07720] Unable to choose an output format for 'invalid'; use a standard extension for the filename or specify the format manually.
[AVFormatContext @ 0x12e006c50] [out#0 @ 0x11df12660] Unable to choose an output format for 'invalid'; use a standard extension for the filename or specify the format manually.
```
IMO it would be better if the output of each process was captured and printed to stdout or stderr in a block, so you can follow the messages from each subprocess.